File: src/emitter/mermaid/theme/nest-palette-index.ts
/**
 * Maps a 1-based subgraph depth to a 0-based palette index by cycling. A
 * palette of length N covers depths 1..N as palette[0]..palette[N-1], then
 * depth N+1 wraps back to palette[0]. Depth must be >= 1; palette must be
 * non-empty.
 */
export function nestPaletteIndex(depth: number, paletteLength: number): number {
  if (paletteLength <= 0) {
    throw new Error("paletteLength must be > 0");
  }
  if (depth < 1) {
    throw new Error("depth must be >= 1");
  }
  return (depth - 1) % paletteLength;
}

File: crates/unsnarl-emitter-mermaid/src/theme/nest_palette_index.rs
//! Maps a 1-based subgraph depth to a 0-based palette index by
//! cycling.
//!
//! Panics on invalid inputs so a mis-wired emitter fails loudly
//! rather than silently picking the wrong color slot.

pub fn nest_palette_index(depth: u32, palette_length: usize) -> usize {
    if palette_length == 0 {
        panic!("paletteLength must be > 0");
    }
    if depth < 1 {
        panic!("depth must be >= 1");
    }
    ((depth as usize) - 1) % palette_length
}

#[cfg(test)]
#[path = "nest_palette_index_test.rs"]
mod nest_palette_index_test;
